embassy-nrf/src/uarte.rs

```diff
@@ -52,188 +55,179 @@ where
+
+        if let Some(pin) = rts.pin_mut() {
+            pin.set_high();
+            pin.conf().write(|w| w.dir().output().drive().h0h1());
+        }
+        r.psel.cts.write(|w| unsafe { w.bits(cts.psel_bits()) });
 
-        Uarte {
-            instance: uarte,
+        if let Some(pin) = cts.pin_mut() {
+            pin.conf().write(|w| w.input().connect().drive().h0h1());
+        }
+        r.psel.rts.write(|w| unsafe { w.bits(rts.psel_bits()) });
+
+        r.baudrate.write(|w| w.baudrate().variant(config.baudrate));
+        r.config.write(|w| w.parity().variant(config.parity));
+
+        // Enable
+        r.enable.write(|w| w.enable().enabled());
+
+        Self {
+            peri: uarte,
             irq,
-            pins,
+            phantom: PhantomData,
         }
     }
+}
 
-    pub fn free(self) -> (T, T::Interrupt, Pins) {
-        // Wait for the peripheral to be disabled from the ISR.
-        while self.instance.enable.read().enable().is_enabled() {}
-        (self.instance, self.irq, self.pins)
-    }
+impl<'d, T: Instance> Drop for Uarte<'d, T> {
+    fn drop(&mut self) {
+        let r = self.peri.regs();
+        r.enable.write(|w| w.enable().disabled());
 
-    fn enable(&mut self) {
-        trace!("enable");
-        self.instance.enable.write(|w| w.enable().enabled());
+        // todo disable pins
     }
+}
 
-    fn tx_started(&self) -> bool {
-        self.instance.events_txstarted.read().bits() != 0
-    }
+impl<'d, T: Instance> Read for Uarte<'d, T> {
+    #[rustfmt::skip]
+    type ReadFuture<'a> where Self: 'a = impl Future<Output = Result<(), Error>> + 'a;
 
-    fn rx_started(&self) -> bool {
-        self.instance.events_rxstarted.read().bits() != 0
-    }
+    fn read<'a>(self: Pin<&'a mut Self>, rx_buffer: &'a mut [u8]) -> Self::ReadFuture<'a> {
+        async move {
+            let this = unsafe { self.get_unchecked_mut() };
 
-    unsafe fn on_irq(_ctx: *mut ()) {
-        let uarte = &*pac::UARTE0::ptr();
+            let ptr = rx_buffer.as_ptr();
+            let len = rx_buffer.len();
+            assert!(len <= EASY_DMA_SIZE);
 
-        let mut try_disable = false;
+            let r = this.peri.regs();
 
-        if uarte.events_endtx.read().bits() != 0 {
-            uarte.events_endtx.reset();
-            trace!("endtx");
-            compiler_fence(Ordering::SeqCst);
+            let drop = OnDrop::new(move || {
+                info!("read drop: stopping");
 
-            if uarte.events_txstarted.read().bits() != 0 {
-                // The ENDTX was signal triggered because DMA has finished.
-                uarte.events_txstarted.reset();
-                try_disable = true;
-            }
+                r.intenclr.write(|w| w.endrx().clear());
+                r.tasks_stoprx.write(|w| unsafe { w.bits(1) });
 
-            T::state().tx_done.signal(());
-        }
+                // TX is stopped almost instantly, spinning is fine.
```

Comment:
RX abort is slow, previously we used `embassy_extras::low_power_wait_until()` to wait until it is over.